From 158d6bfc8ae7ee803ace6c7387c517580a2e9cc2 Mon Sep 17 00:00:00 2001
From: Armin Wolf <W_Armin@gmx.de>
Date: Thu, 5 Jan 2023 23:51:07 +0100
Subject: [PATCH 36/69] hwmon: (ftsteutates) Add support for fanX_fault
 attributes

The driver knows internally when a fan is not connected,
but does not export this knowledge to userspace. Use the
standard fanX_fault attributes to notify userspace if a
fan is not connected.

Tested on a Fujitsu DS3401-B1.

Signed-off-by: Armin Wolf <W_Armin@gmx.de>
Link: https://lore.kernel.org/r/20230105225107.58308-4-W_Armin@gmx.de
Signed-off-by: Guenter Roeck <linux@roeck-us.net>
---
 drivers/hwmon/ftsteutates.c | 21 +++++++++++++--------
 1 file changed, 13 insertions(+), 8 deletions(-)

diff --git a/drivers/hwmon/ftsteutates.c b/drivers/hwmon/ftsteutates.c
index 0d8ab9425..25afd9167 100644
--- a/drivers/hwmon/ftsteutates.c
+++ b/drivers/hwmon/ftsteutates.c
@@ -356,6 +356,7 @@ static umode_t fts_is_visible(const void *devdata, enum hwmon_sensor_types type,
 	case hwmon_fan:
 		switch (attr) {
 		case hwmon_fan_input:
+		case hwmon_fan_fault:
 			return 0444;
 		case hwmon_fan_alarm:
 			return 0644;
@@ -411,6 +412,10 @@ static int fts_read(struct device *dev, enum hwmon_sensor_types type, u32 attr,
 		case hwmon_fan_alarm:
 			*val = !!(data->fan_alarm & BIT(channel));
 
+			return 0;
+		case hwmon_fan_fault:
+			*val = !(data->fan_present & BIT(channel));
+
 			return 0;
 		default:
 			break;
@@ -536,14 +541,14 @@ static const struct hwmon_channel_info *fts_info[] = {
 			   HWMON_T_INPUT | HWMON_T_ALARM | HWMON_T_FAULT
 			   ),
 	HWMON_CHANNEL_INFO(fan,
-			   HWMON_F_INPUT | HWMON_F_ALARM,
-			   HWMON_F_INPUT | HWMON_F_ALARM,
-			   HWMON_F_INPUT | HWMON_F_ALARM,
-			   HWMON_F_INPUT | HWMON_F_ALARM,
-			   HWMON_F_INPUT | HWMON_F_ALARM,
-			   HWMON_F_INPUT | HWMON_F_ALARM,
-			   HWMON_F_INPUT | HWMON_F_ALARM,
-			   HWMON_F_INPUT | HWMON_F_ALARM
+			   HWMON_F_INPUT | HWMON_F_ALARM | HWMON_F_FAULT,
+			   HWMON_F_INPUT | HWMON_F_ALARM | HWMON_F_FAULT,
+			   HWMON_F_INPUT | HWMON_F_ALARM | HWMON_F_FAULT,
+			   HWMON_F_INPUT | HWMON_F_ALARM | HWMON_F_FAULT,
+			   HWMON_F_INPUT | HWMON_F_ALARM | HWMON_F_FAULT,
+			   HWMON_F_INPUT | HWMON_F_ALARM | HWMON_F_FAULT,
+			   HWMON_F_INPUT | HWMON_F_ALARM | HWMON_F_FAULT,
+			   HWMON_F_INPUT | HWMON_F_ALARM | HWMON_F_FAULT
 			   ),
 	HWMON_CHANNEL_INFO(pwm,
 			   HWMON_PWM_AUTO_CHANNELS_TEMP,
-- 
2.39.2.501.gd9d677b2d8

